File: automaton/src/interagent/scaling/cluster.ts
/**
 * 水平扩展支持模块
 * 实现多实例水平扩展能力
 *
 * @module interagent.scaling.cluster
 * @version 1.0.0
 */

import { EventEmitter } from "events";

// ============================================================================
// Types
// ============================================================================

/**
 * 实例状态
 */
export type InstanceStatus = "healthy" | "degraded" | "unhealthy";

/**
 * 负载均衡策略
 */
export type LoadBalanceStrategy = "round-robin" | "least-load" | "random";

/**
 * 实例信息
 */
export interface InstanceInfo {
  /** 实例 ID */
  id: string;
  /** 实例 URL */
  url: string;
  /** 实例状态 */
  status: InstanceStatus;
  /** 当前负载 (0-1) */
  load: number;
  /** 最后心跳时间 */
  lastHeartbeat: Date;
  /** 元数据 */
  metadata?: Record<string, unknown>;
}

/**
 * 实例注册选项
 */
export interface RegisterOptions {
  /** 超时时间 (毫秒) */
  heartbeatTimeout?: number;
  /** 元数据 */
  metadata?: Record<string, unknown>;
}

/**
 * 路由结果
 */
export interface RouteResult {
  /** 选中的实例 */
  instance: InstanceInfo;
  /** 路由策略 */
  strategy: LoadBalanceStrategy;
}

/**
 * 集群统计
 */
export interface ClusterStats {
  /** 总实例数 */
  totalInstances: number;
  /** 健康实例数 */
  healthyInstances: number;
  /** 降级实例数 */
  degradedInstances: number;
  /** 不健康实例数 */
  unhealthyInstances: number;
  /** 总负载 */
  totalLoad: number;
  /** 平均负载 */
  averageLoad: number;
}

// ============================================================================
// Constants
// ============================================================================

const DEFAULT_HEARTBEAT_INTERVAL = 30000; // 30 秒
const DEFAULT_HEARTBEAT_TIMEOUT = 60000; // 60 秒

// ============================================================================
// InstanceRegistry Class
// ============================================================================

/**
 * 实例注册表
 *
 * 管理集群中的所有实例，提供注册、发现和健康检查功能
 *
 * @example
 * ```typescript
 * const registry = new InstanceRegistry();
 *
 * // 注册实例
 * registry.register({
 *   id: 'instance-1',
 *   url: 'http://localhost:3001',
 *   status: 'healthy',
 *   load: 0.5,
 * });
 *
 * // 获取健康实例
 * const healthy = registry.getHealthyInstances();
 *
 * // 选择实例
 * const selected = registry.selectInstance('least-load');
 * ```
 */
export class InstanceRegistry extends EventEmitter {
  private instances: Map<string, InstanceInfo> = new Map();
  private heartbeatTimeout: number;
  private roundRobinIndex = 0;

  constructor(options: { heartbeatTimeout?: number } = {}) {
    super();
    this.heartbeatTimeout = options.heartbeatTimeout || DEFAULT_HEARTBEAT_TIMEOUT;
  }

  /**
   * 注册实例
   */
  register(instance: Omit<InstanceInfo, "lastHeartbeat">, options?: RegisterOptions): void {
    const info: InstanceInfo = {
      ...instance,
      lastHeartbeat: new Date(),
      metadata: options?.metadata || instance.metadata,
    };

    const isNew = !this.instances.has(instance.id);
    this.instances.set(instance.id, info);

    if (isNew) {
      this.emit("registered", info);
    } else {
      this.emit("updated", info);
    }
  }

  /**
   * 注销实例
   */
  deregister(instanceId: string): boolean {
    const instance = this.instances.get(instanceId);
    if (instance) {
      this.instances.delete(instanceId);
      this.emit("deregistered", instance);
      return true;
    }
    return false;
  }

  /**
   * 更新心跳
   */
  heartbeat(instanceId: string, load: number, status?: InstanceStatus): boolean {
    const instance = this.instances.get(instanceId);
    if (instance) {
      instance.lastHeartbeat = new Date();
      instance.load = Math.max(0, Math.min(1, load));
      if (status) {
        instance.status = status;
      }
      this.emit("heartbeat", instance);
      return true;
    }
    return false;
  }

  /**
   * 获取实例
   */
  getInstance(instanceId: string): InstanceInfo | undefined {
    return this.instances.get(instanceId);
  }

  /**
   * 获取所有实例
   */
  getAllInstances(): InstanceInfo[] {
    return Array.from(this.instances.values());
  }

  /**
   * 获取健康实例
   */
  getHealthyInstances(): InstanceInfo[] {
    const now = Date.now();
    return Array.from(this.instances.values()).filter(
      (i) =>
        (i.status === "healthy" || i.status === "degraded") &&
        now - i.lastHeartbeat.getTime() < this.heartbeatTimeout
    );
  }

  /**
   * 按状态获取实例
   */
  getInstancesByStatus(status: InstanceStatus): InstanceInfo[] {
    return Array.from(this.instances.values()).filter((i) => i.status === status);
  }

  /**
   * 选择实例
   */
  selectInstance(strategy: LoadBalanceStrategy = "least-load"): InstanceInfo | null {
    const healthy = this.getHealthyInstances();
    if (healthy.length === 0) return null;

    switch (strategy) {
      case "least-load":
        return healthy.reduce((a, b) => (a.load < b.load ? a : b));

      case "round-robin":
        this.roundRobinIndex = (this.roundRobinIndex + 1) % healthy.length;
        return healthy[this.roundRobinIndex];

      case "random":
        return healthy[Math.floor(Math.random() * healthy.length)];

      default:
        return healthy[0];
    }
  }

  /**
   * 检查实例健康状态
   */
  checkHealth(): { healthy: string[]; unhealthy: string[] } {
    const now = Date.now();
    const healthy: string[] = [];
    const unhealthy: string[] = [];

    for (const [id, instance] of this.instances) {
      if (
        instance.status === "healthy" &&
        now - instance.lastHeartbeat.getTime() < this.heartbeatTimeout
      ) {
        healthy.push(id);
      } else {
        unhealthy.push(id);
      }
    }

    return { healthy, unhealthy };
  }

  /**
   * 获取统计信息
   */
  getStats(): ClusterStats {
    const instances = Array.from(this.instances.values());
    const totalInstances = instances.length;
    const healthyInstances = instances.filter((i) => i.status === "healthy").length;
    const degradedInstances = instances.filter((i) => i.status === "degraded").length;
    const unhealthyInstances = instances.filter((i) => i.status === "unhealthy").length;
    const totalLoad = instances.reduce((sum, i) => sum + i.load, 0);
    const averageLoad = totalInstances > 0 ? totalLoad / totalInstances : 0;

    return {
      totalInstances,
      healthyInstances,
      degradedInstances,
      unhealthyInstances,
      totalLoad,
      averageLoad,
    };
  }

  /**
   * 清空所有实例
   */
  clear(): void {
    this.instances.clear();
    this.roundRobinIndex = 0;
    this.emit("cleared");
  }

  /**
   * 获取实例数量
   */
  get size(): number {
    return this.instances.size;
  }
}

// ============================================================================
// LoadBalancer Class
// ============================================================================

/**
 * 负载均衡器
 *
 * 根据策略将请求路由到合适的实例
 *
 * @example
 * ```typescript
 * const registry = new InstanceRegistry();
 * const loadBalancer = new LoadBalancer(registry);
 *
 * // 路由请求
 * const result = loadBalancer.selectInstance();
 * if (result) {
 *   console.log(`Routing to ${result.instance.url}`);
 * }
 * ```
 */
export class LoadBalancer {
  private registry: InstanceRegistry;
  private strategy: LoadBalanceStrategy;
  private sessionAffinity: Map<string, string> = new Map();

  constructor(registry: InstanceRegistry, strategy: LoadBalanceStrategy = "least-load") {
    this.registry = registry;
    this.strategy = strategy;
  }

  /**
   * 设置负载均衡策略
   */
  setStrategy(strategy: LoadBalanceStrategy): void {
    this.strategy = strategy;
  }

  /**
   * 选择实例
   */
  selectInstance(sessionId?: string): RouteResult | null {
    // 会话亲和性检查
    if (sessionId) {
      const affinityInstance = this.sessionAffinity.get(sessionId);
      if (affinityInstance) {
        const instance = this.registry.getInstance(affinityInstance);
        if (instance && instance.status !== "unhealthy") {
          return { instance, strategy: "round-robin" };
        }
        // 实例不健康，移除亲和性
        this.sessionAffinity.delete(sessionId);
      }
    }

    const instance = this.registry.selectInstance(this.strategy);
    if (!instance) return null;

    // 设置会话亲和性
    if (sessionId) {
      this.sessionAffinity.set(sessionId, instance.id);
    }

    return { instance, strategy: this.strategy };
  }

  /**
   * 路由请求
   */
  async routeRequest<T>(
    request: T,
    executor: (instance: InstanceInfo, request: T) => Promise<Response>
  ): Promise<{ response: Response; instance: InstanceInfo }> {
    const result = this.selectInstance();
    if (!result) {
      throw new Error("No healthy instances available");
    }

    try {
      const response = await executor(result.instance, request);
      return { response, instance: result.instance };
    } catch (error) {
      // 故障转移：尝试其他实例
      const retryResult = this.registry.selectInstance("random");
      if (retryResult && retryResult.id !== result.instance.id) {
        const response = await executor(retryResult, request);
        return { response, instance: retryResult };
      }
      throw error;
    }
  }

  /**
   * 清除会话亲和性
   */
  clearSessionAffinity(sessionId?: string): void {
    if (sessionId) {
      this.sessionAffinity.delete(sessionId);
    } else {
      this.sessionAffinity.clear();
    }
  }

  /**
   * 获取注册表
   */
  getRegistry(): InstanceRegistry {
    return this.registry;
  }
}

// ============================================================================
// Default Export
// ============================================================================

export default InstanceRegistry;

File: nanobot/nanobot/interagent/scaling/cluster.py
"""
水平扩展支持模块
实现多实例水平扩展能力

@module interagent.scaling.cluster
@version 1.0.0
"""

from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional, Callable, TypeVar, Generic
from collections.abc import Awaitable
import asyncio
import random


# ============================================================================
# Types
# ============================================================================


class InstanceStatus(str, Enum):
    """实例状态"""

    HEALTHY = "healthy"
    DEGRADED = "degraded"
    UNHEALTHY = "unhealthy"


class LoadBalanceStrategy(str, Enum):
    """负载均衡策略"""

    ROUND_ROBIN = "round-robin"
    LEAST_LOAD = "least-load"
    RANDOM = "random"


@dataclass
class InstanceInfo:
    """实例信息"""

    id: str
    """实例 ID"""

    url: str
    """实例 URL"""

    status: InstanceStatus = InstanceStatus.HEALTHY
    """实例状态"""

    load: float = 0.0
    """当前负载 (0-1)"""

    last_heartbeat: datetime = field(default_factory=datetime.now)
    """最后心跳时间"""

    metadata: Dict[str, Any] = field(default_factory=dict)
    """元数据"""


@dataclass
class ClusterStats:
    """集群统计"""

    total_instances: int = 0
    """总实例数"""

    healthy_instances: int = 0
    """健康实例数"""

    degraded_instances: int = 0
    """降级实例数"""

    unhealthy_instances: int = 0
    """不健康实例数"""

    total_load: float = 0.0
    """总负载"""

    average_load: float = 0.0
    """平均负载"""


T = TypeVar("T")


# ============================================================================
# InstanceRegistry Class
# ============================================================================


class InstanceRegistry:
    """
    实例注册表

    管理集群中的所有实例，提供注册、发现和健康检查功能

    Example:
        registry = InstanceRegistry()

        # 注册实例
        registry.register(InstanceInfo(
            id='instance-1',
            url='http://localhost:3001',
            status=InstanceStatus.HEALTHY,
            load=0.5,
        ))

        # 获取健康实例
        healthy = registry.get_healthy_instances()

        # 选择实例
        selected = registry.select_instance(LoadBalanceStrategy.LEAST_LOAD)
    """

    def __init__(self, heartbeat_timeout: float = 60.0):
        """
        初始化实例注册表

        Args:
            heartbeat_timeout: 心跳超时时间 (秒)
        """
        self._instances: Dict[str, InstanceInfo] = {}
        self._heartbeat_timeout = heartbeat_timeout
        self._round_robin_index = 0
        self._listeners: Dict[str, List[Callable]] = {
            "registered": [],
            "deregistered": [],
            "updated": [],
            "heartbeat": [],
        }

    def on(self, event: str, callback: Callable) -> None:
        """注册事件监听器"""
        if event in self._listeners:
            self._listeners[event].append(callback)

    def off(self, event: str, callback: Callable) -> None:
        """移除事件监听器"""
        if event in self._listeners and callback in self._listeners[event]:
            self._listeners[event].remove(callback)

    def _emit(self, event: str, data: Any) -> None:
        """触发事件"""
        for callback in self._listeners.get(event, []):
            try:
                callback(data)
            except Exception:
                pass

    def register(
        self,
        instance: InstanceInfo,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> None:
        """
        注册实例

        Args:
            instance: 实例信息
            metadata: 可选的元数据
        """
        instance.last_heartbeat = datetime.now()
        if metadata:
            instance.metadata = {**instance.metadata, **metadata}

        is_new = instance.id not in self._instances
        self._instances[instance.id] = instance

        if is_new:
            self._emit("registered", instance)
        else:
            self._emit("updated", instance)

    def deregister(self, instance_id: str) -> bool:
        """
        注销实例

        Args:
            instance_id: 实例 ID

        Returns:
            是否成功注销
        """
        instance = self._instances.get(instance_id)
        if instance:
            del self._instances[instance_id]
            self._emit("deregistered", instance)
            return True
        return False

    def heartbeat(
        self,
        instance_id: str,
        load: float,
        status: Optional[InstanceStatus] = None,
    ) -> bool:
        """
        更新心跳

        Args:
            instance_id: 实例 ID
            load: 当前负载
            status: 可选的状态更新

        Returns:
            是否成功更新
        """
        instance = self._instances.get(instance_id)
        if instance:
            instance.last_heartbeat = datetime.now()
            instance.load = max(0.0, min(1.0, load))
            if status:
                instance.status = status
            self._emit("heartbeat", instance)
            return True
        return False

    def get_instance(self, instance_id: str) -> Optional[InstanceInfo]:
        """获取实例"""
        return self._instances.get(instance_id)

    def get_all_instances(self) -> List[InstanceInfo]:
        """获取所有实例"""
        return list(self._instances.values())

    def get_healthy_instances(self) -> List[InstanceInfo]:
        """获取健康实例"""
        now = datetime.now()
        timeout_seconds = self._heartbeat_timeout

        return [
            i
            for i in self._instances.values()
            if i.status in (InstanceStatus.HEALTHY, InstanceStatus.DEGRADED)
            and (now - i.last_heartbeat).total_seconds() < timeout_seconds
        ]

    def get_instances_by_status(self, status: InstanceStatus) -> List[InstanceInfo]:
        """按状态获取实例"""
        return [i for i in self._instances.values() if i.status == status]

    def select_instance(
        self, strategy: LoadBalanceStrategy = LoadBalanceStrategy.LEAST_LOAD
    ) -> Optional[InstanceInfo]:
        """
        选择实例

        Args:
            strategy: 负载均衡策略

        Returns:
            选中的实例或 None
        """
        healthy = self.get_healthy_instances()
        if not healthy:
            return None

        if strategy == LoadBalanceStrategy.LEAST_LOAD:
            return min(healthy, key=lambda i: i.load)

        if strategy == LoadBalanceStrategy.ROUND_ROBIN:
            self._round_robin_index = (self._round_robin_index + 1) % len(healthy)
            return healthy[self._round_robin_index]

        # RANDOM
        return random.choice(healthy)

    def check_health(self) -> Dict[str, List[str]]:
        """
        检查实例健康状态

        Returns:
            包含 healthy 和 unhealthy 列表的字典
        """
        now = datetime.now()
        timeout_seconds = self._heartbeat_timeout

        healthy = []
        unhealthy = []

        for instance_id, instance in self._instances.items():
            if (
                instance.status == InstanceStatus.HEALTHY
                and (now - instance.last_heartbeat).total_seconds() < timeout_seconds
            ):
                healthy.append(instance_id)
            else:
                unhealthy.append(instance_id)

        return {"healthy": healthy, "unhealthy": unhealthy}

    def get_stats(self) -> ClusterStats:
        """获取统计信息"""
        instances = list(self._instances.values())
        total = len(instances)

        if total == 0:
            return ClusterStats()

        healthy = sum(1 for i in instances if i.status == InstanceStatus.HEALTHY)
        degraded = sum(1 for i in instances if i.status == InstanceStatus.DEGRADED)
        unhealthy = sum(1 for i in instances if i.status == InstanceStatus.UNHEALTHY)
        total_load = sum(i.load for i in instances)

        return ClusterStats(
            total_instances=total,
            healthy_instances=healthy,
            degraded_instances=degraded,
            unhealthy_instances=unhealthy,
            total_load=total_load,
            average_load=total_load / total,
        )

    def clear(self) -> None:
        """清空所有实例"""
        self._instances.clear()
        self._round_robin_index = 0

    @property
    def size(self) -> int:
        """获取实例数量"""
        return len(self._instances)


# ============================================================================
# LoadBalancer Class
# ============================================================================


class LoadBalancer:
    """
    负载均衡器

    根据策略将请求路由到合适的实例

    Example:
        registry = InstanceRegistry()
        load_balancer = LoadBalancer(registry)

        # 选择实例
        result = load_balancer.select_instance()
        if result:
            print(f"Routing to {result.url}")
    """

    def __init__(
        self,
        registry: InstanceRegistry,
        strategy: LoadBalanceStrategy = LoadBalanceStrategy.LEAST_LOAD,
    ):
        """
        初始化负载均衡器

        Args:
            registry: 实例注册表
            strategy: 负载均衡策略
        """
        self._registry = registry
        self._strategy = strategy
        self._session_affinity: Dict[str, str] = {}

    def set_strategy(self, strategy: LoadBalanceStrategy) -> None:
        """设置负载均衡策略"""
        self._strategy = strategy

    def select_instance(self, session_id: Optional[str] = None) -> Optional[InstanceInfo]:
        """
        选择实例

        Args:
            session_id: 可选的会话 ID，用于会话亲和性

        Returns:
            选中的实例或 None
        """
        # 会话亲和性检查
        if session_id:
            affinity_instance_id = self._session_affinity.get(session_id)
            if affinity_instance_id:
                instance = self._registry.get_instance(affinity_instance_id)
                if instance and instance.status != InstanceStatus.UNHEALTHY:
                    return instance
                # 实例不健康，移除亲和性
                del self._session_affinity[session_id]

        instance = self._registry.select_instance(self._strategy)
        if not instance:
            return None

        # 设置会话亲和性
        if session_id:
            self._session_affinity[session_id] = instance.id

        return instance

    async def route_request(
        self,
        request: T,
        executor: Callable[[InstanceInfo, T], Awaitable[Any]],
        session_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        路由请求

        Args:
            request: 请求对象
            executor: 执行函数
            session_id: 可选的会话 ID

        Returns:
            包含 response 和 instance 的字典

        Raises:
            RuntimeError: 没有可用的健康实例
        """
        instance = self.select_instance(session_id)
        if not instance:
            raise RuntimeError("No healthy instances available")

        try:
            response = await executor(instance, request)
            return {"response": response, "instance": instance}
        except Exception:
            # 故障转移：尝试其他实例
            retry_instance = self._registry.select_instance(LoadBalanceStrategy.RANDOM)
            if retry_instance and retry_instance.id != instance.id:
                response = await executor(retry_instance, request)
                return {"response": response, "instance": retry_instance}
            raise

    def clear_session_affinity(self, session_id: Optional[str] = None) -> None:
        """清除会话亲和性"""
        if session_id:
            self._session_affinity.pop(session_id, None)
        else:
            self._session_affinity.clear()

    @property
    def registry(self) -> InstanceRegistry:
        """获取注册表"""
        return self._registry
